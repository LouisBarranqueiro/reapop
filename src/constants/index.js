export const DEFAULT_STATUS = 'default';
export const INFO_STATUS = 'info';
export const SUCCESS_STATUS = 'success';
export const WARNING_STATUS = 'warning';
export const ERROR_STATUS = 'error';

export const STATUS = {
  default: DEFAULT_STATUS,
  info: INFO_STATUS,
  success: SUCCESS_STATUS,
  warning: WARNING_STATUS,
  error: ERROR_STATUS
};

export const TOP = 't';
export const TOP_CENTER = 'tc';
export const TOP_LEFT_POSITION = 'tl';
export const TOP_RIGHT_POSITION = 'tr';
export const BOTTOM = 'b';
export const BOTTOM_CENTER = 'bc';
export const BOTTOM_LEFT_POSITION = 'bl';
export const BOTTOM_RIGHT_POSITION = 'br';

export const POSITIONS = {
  top: TOP,
  topCenter: TOP_CENTER,
  topLeft: TOP_LEFT_POSITION,
  topRight: TOP_RIGHT_POSITION,
  bottom: BOTTOM,
  bottomCenter: BOTTOM_CENTER,
  bottomLeft: BOTTOM_LEFT_POSITION,
  bottomRight: BOTTOM_RIGHT_POSITION
};

// default value for notifications
export const DEFAULT_NOTIFICATION = {
  status: DEFAULT_STATUS,
  position: TOP_RIGHT_POSITION,
  dismissible: true,
  dismissAfter: 5000,
  allowHTML: false,
  closeButton: false
};
