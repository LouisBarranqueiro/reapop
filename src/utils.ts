type TimeoutId = ReturnType<typeof setTimeout>

export class Timer {
    private timerId: TimeoutId | undefined
    private start: Date | undefined
    private callback: (...args: any[]) => any
    remainingTime: number

    constructor(delay: number, callback: (...args: any[]) => any) {
        this.remainingTime = delay
        this.callback = callback
    }

    pause() {
        clearTimeout(<TimeoutId>this.timerId)
        this.remainingTime -= new Date().getTime() - (<Date>this.start).getTime()
    }

    resume() {
        this.start = new Date()
        clearTimeout(<TimeoutId>this.timerId)
        this.timerId = setTimeout(this.callback, this.remainingTime)
    }
}

export const clone = <T extends unknown>(origObject: T): T => {
    if (typeof origObject !== 'object' || origObject === null) {
        return origObject
    }

    if (Array.isArray(origObject)) {
        return origObject.map((value) => clone(value)) as T
    }

    const newObject: Record<string, unknown> = {}
    for (const key in origObject) {
        newObject[key] = clone((origObject as Record<string, unknown>)[key])
    }

    return newObject as T
}

export const isObject = (value: unknown) => !!value && (value as any).constructor === Object
