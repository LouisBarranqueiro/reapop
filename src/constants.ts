import {NewNotification, Notification, Position} from './reducers/notifications/types'

export type NotificationConfig = {
    defaultProps: Partial<Notification>
    customizeNotification: ((notification: NewNotification) => void) | null
    generateId: ((notification: NewNotification) => string) | null
}

export const CONFIG: NotificationConfig = {
    defaultProps: {},
    customizeNotification: null,
    generateId: null,
}

export const STATUSES: {
    none: 'none'
    info: 'info'
    success: 'success'
    loading: 'loading'
    warning: 'warning'
    error: 'error'
} = Object.freeze({
    none: 'none',
    info: 'info',
    success: 'success',
    loading: 'loading',
    warning: 'warning',
    error: 'error',
})

export const POSITIONS: {
    topCenter: 'top-center'
    topLeft: 'top-left'
    topRight: 'top-right'
    bottomCenter: 'bottom-center'
    bottomLeft: 'bottom-left'
    bottomRight: 'bottom-right'
} = Object.freeze({
    topCenter: 'top-center',
    topLeft: 'top-left',
    topRight: 'top-right',
    bottomCenter: 'bottom-center',
    bottomLeft: 'bottom-left',
    bottomRight: 'bottom-right',
})

const classnamePrefix = 'reapop__'

export const classnames = {
    container: (position: Position, singleContainer: boolean) => {
        const classes = [`${classnamePrefix}container`, `${classnamePrefix}container--${position}`]
        if (singleContainer) {
            classes.push(`${classnamePrefix}container--single`)
        }
        return classes
    },
    notification: (notification: Notification) => [
        `${classnamePrefix}notification`,
        `${classnamePrefix}notification--${notification.status}`,
        `${classnamePrefix}notification--buttons-${notification.buttons.length}`,
        notification.dismissible ? `${classnamePrefix}notification--dismissible` : '',
    ],
    notificationIcon: `${classnamePrefix}notification-icon`,
    notificationImageContainer: `${classnamePrefix}notification-image-container`,
    notificationImage: `${classnamePrefix}notification-image`,
    notificationDismissIcon: `${classnamePrefix}notification-dismiss-icon`,
    notificationMeta: `${classnamePrefix}notification-meta`,
    notificationTitle: `${classnamePrefix}notification-title`,
    notificationMessage: `${classnamePrefix}notification-message`,
    notificationButtons: `${classnamePrefix}notification-buttons`,
    notificationButton: `${classnamePrefix}notification-button`,
    notificationButtonText: `${classnamePrefix}notification-button-text`,
}
