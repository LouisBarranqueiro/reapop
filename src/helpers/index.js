import {STATUS} from '../constants';

/**
 * Convert status in a understandable status for the Notification component
 * @param {String|Number} status
 * @returns {String} status an understandable status
 */
export function convertStatus(status) {
  const reHttpStatusCode = /^\d{3}$/;
  // convert HTTP status code
  if (reHttpStatusCode.test(status)) {
    switch (true) {
      case /^1/.test(status):
        return STATUS.info;
      case /^2/.test(status):
        return STATUS.success;
      case /^(4|5)/.test(status):
        return STATUS.error;
    }
  }
  return status;
}

/**
 * Create a Timer
 * @param {Function} callback
 * @param {Number} delay
 * @constructor
 */
export function Timer(delay, callback) {
  let timerId;
  let start;
  let remaining = delay;

  this.pause = () => {
    clearTimeout(timerId);
    remaining -= new Date() - start;
  };
  this.resume = () => {
    start = new Date();
    clearTimeout(timerId);
    timerId = setTimeout(callback, remaining);
  };

  this.getTimeRemaining = () => {
    return remaining;
  };
}

/**
 * Treat data of a notification
 * @param {Object} notification
 * @returns {Object} a notification
 */
export function treatNotification(notification) {
  if (notification.dismissAfter) {
    notification.dismissAfter = parseInt(notification.dismissAfter);
  }
  if (notification.image) {
    notification.status = STATUS.default;
  }
  else {
    notification.status = convertStatus(notification.status);
  }
  if (!notification.buttons) {
    notification.buttons = [];
  }
  return notification;
}

/**
 * Preload an image
 * @param {String} url url of image to load
 * @param {Function} onload Function called when image is loaded
 * @returns {void}
 */
export function preloadImage(url, onload) {
  var image = new Image();
  image.src = url;
  image.onload = onload;
  return image;
}

/**
 * Return values of an Object in an Array
 * @param {Object} obj
 * @returns {Array}
 */
export function mapObjectValues(obj) {
  var array = [];

  for (let key in obj) {
    if (obj.hasOwnProperty(key)) {
      array.push(obj[key]);
    }
  }
  return array;
}
