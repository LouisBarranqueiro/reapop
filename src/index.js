import NotificationsSystem from './components/NotificationsSystem';
import {STATUS, POSITIONS, DEFAULT_NOTIFICATION} from './constants';
import reducer, {
  actions,
  types,
  addNotification,
  updateNotification,
  removeNotification,
  removeNotifications
} from './store/notifications';

export {
  // constants
  // default value for notifications
  DEFAULT_NOTIFICATION,
  // all available status
  STATUS,
  // all available positions
  POSITIONS,
  // all action creators
  actions,
  // actions types
  types,
  // reducer
  reducer,
  // action creators separately
  addNotification,
  updateNotification,
  removeNotification,
  removeNotifications
};
// NotificationsSystem React component
export default NotificationsSystem;
