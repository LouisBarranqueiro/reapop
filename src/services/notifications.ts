import {CONFIG, NotificationConfig, STATUSES} from '../constants'
import {NewNotification, Notification} from '../reducers/notifications/types'
import {clone} from '../utils'

export const setUpNotifications = (props: Partial<NotificationConfig>) => {
    for (const key in props) {
        const propName = key as keyof NotificationConfig
        CONFIG[propName] = props[propName] as any
    }
}

export const resetNotificationsConfig = () => {
    CONFIG.defaultProps = {}
    CONFIG.customizeNotification = null
    CONFIG.generateId = null
}

export const generateId = () => Math.random().toString(36).slice(2, 7)

export const prepareNotification = (notification: NewNotification): Notification => {
    const notif = clone(notification)
    const defaultProps: Partial<Notification> = {
        ...CONFIG.defaultProps,
        status: STATUSES.none,
        buttons: [],
    }

    if (notif.id === undefined) {
        notif.id = CONFIG.generateId ? CONFIG.generateId(notification) : generateId()
    }

    Object.entries(defaultProps).map(([prop, defaultValue]) => {
        if (notif[prop] === undefined) {
            notif[prop] = defaultValue
        }
    })

    if (typeof CONFIG.customizeNotification === 'function') {
        CONFIG.customizeNotification(notif)
    }
    return notif as Notification
}
