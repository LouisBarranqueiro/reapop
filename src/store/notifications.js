import {treatNotification, preloadImage} from '../helpers';
import {DEFAULT_NOTIFICATION} from '../constants';

// An array to store notifications object
const INITIAL_STATE = [];
// Action types
const ADD_NOTIFICATION = 'ADD_NOTIFICATION';
const UPDATE_NOTIFICATION = 'UPDATE_NOTIFICATION';
const REMOVE_NOTIFICATION = 'REMOVE_NOTIFICATION';
const REMOVE_NOTIFICATIONS = 'REMOVE_NOTIFICATIONS';

/**
 * Add a notification (thunk action creator)
 *
 * We use a thunk here to create an ADD_NOTIFICATION action
 * and only return the notification object.
 * @param {Object} notification
 * @returns {Object} notification
 */
export const addNotification = (notification) => (dispatch) => {
  if (!notification.id) {
    notification.id = new Date().getTime();
  }
  notification = treatNotification(notification);
  // if there is an image, we preload it
  // and add notification when image is loaded
  if (notification.image) {
    preloadImage(notification.image, dispatch.bind(this, _addNotification(notification)));
  }
  else {
    dispatch(_addNotification(notification));
  }
  return notification;
};

/**
 * Add a notification (action creator)
 *
 * @param {Object} notification
 * @returns {{type: string, payload: {Object}}}
 * @private
 */
function _addNotification(notification) {
  return {
    type: ADD_NOTIFICATION,
    payload: notification
  };
}

/**
 * Update a notification (thunk action creator)
 *
 * We use a thunk here to create an UPDATE_NOTIFICATION action
 * and only return the notification object.
 * @param {Object} notification
 * @returns {Object} notification
 */
export const updateNotification = (notification) => (dispatch, getState) => {
  if (!notification.id) {
    throw new Error('A notification must have an `id` property to be updated');
  }

  const notifications = getState().notifications;
  const index = notifications.findIndex((oldNotification) => oldNotification.id === notification.id);
  const currNotification = notifications[index];

  notification = treatNotification(notification);

  // if image is different, then we preload it
  // and update notification when image is loaded
  if (notification.image && (!currNotification.image || (currNotification.image &&
    notification.image !== currNotification.image))) {
    preloadImage(notification.image, dispatch.bind(this, _updateNotification(notification)));
  }
  else {
    dispatch(_updateNotification(notification));
  }
  return notification;
};

/**
 * Update a notification (action creator)
 *
 * @param {Object} notification
 * @returns {{type: string, payload: {Object}}}
 * @private
 */
function _updateNotification(notification) {
  return {
    type: UPDATE_NOTIFICATION,
    payload: notification
  };
}

/**
 * Remove a notification (action creator)
 *
 * @param {Object} notification
 * @returns {{type: string, payload: {Object}}}
 */
export function removeNotification(notification) {
  return {
    type: REMOVE_NOTIFICATION,
    payload: notification
  };
}

/**
 * Remove all notifications (action creator)
 *
 * @returns {{type: string}}
 */
export function removeNotifications() {
  return {
    type: REMOVE_NOTIFICATIONS
  };
}

// Action creators
export const actions = {
  addNotification: addNotification,
  updateNotification: updateNotification,
  removeNotification: removeNotification,
  removeNotifications: removeNotifications
};

// Actions types
export const types = {
  ADD_NOTIFICATION: ADD_NOTIFICATION,
  UPDATE_NOTIFICATION: UPDATE_NOTIFICATION,
  REMOVE_NOTIFICATION: REMOVE_NOTIFICATION,
  REMOVE_NOTIFICATIONS: REMOVE_NOTIFICATIONS
};

// Reducers
export default (defaultNotification = DEFAULT_NOTIFICATION) => {
  return (state = INITIAL_STATE, {type, payload}) => {
    switch (type) {
      case ADD_NOTIFICATION:
        const notification = Object.assign({}, defaultNotification, payload);
        return [...state, notification];
      case UPDATE_NOTIFICATION:
        // get index of the notification
        const index = state.findIndex((notification) => notification.id === payload.id);
        // replace the old notification by the new one
        state[index] = Object.assign({}, defaultNotification, payload);
        return [...state];
      case REMOVE_NOTIFICATION:
        return state.filter((notification) => notification.id !== payload);
      case REMOVE_NOTIFICATIONS:
        return [];
      default:
        return state;
    }
  };
};
