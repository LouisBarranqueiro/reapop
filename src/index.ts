import NotificationsSystem from './components/NotificationsSystem'
import {NotificationsProvider} from './components/NotificationsProvider'
import {STATUSES, POSITIONS} from './constants'
import {notify, dismissNotifications, dismissNotification} from './reducers/notifications/actions'
import FadeTransition from './components/FadeTransition'
import SlideTransition from './components/SlideTransition'
import GrowTransition from './components/GrowTransition'
import reducer from './reducers/notifications/reducer'
import {useNotifications} from './hooks/useNotifications'
import baseTheme from './themes/base'
import atalhoTheme from './themes/atalho'
import wyboTheme from './themes/wybo'
import bootstrapTheme from './themes/bootstrap'

import {Theme} from './themes/types'
import {setUpNotifications} from './services/notifications'

export {
    // themes
    baseTheme,
    atalhoTheme,
    wyboTheme,
    bootstrapTheme,
    // utils
    setUpNotifications,
    useNotifications,
    // constants
    STATUSES,
    POSITIONS,
    //reducers
    reducer,
    // actions
    notify,
    dismissNotifications,
    dismissNotification,
    // components
    FadeTransition,
    SlideTransition,
    GrowTransition,
    NotificationsProvider,
}

// types
export * from './reducers/notifications/types'
export type {Theme}

export default NotificationsSystem
