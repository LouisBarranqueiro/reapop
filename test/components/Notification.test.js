import React from 'react';
import {mount, shallow} from 'enzyme';
import {Provider} from 'react-redux';
import {Timer} from '../../src/helpers';
import ConnectNotification, {Notification} from '../../src/components/Notification';
import theme from 'reapop-theme-wybo';
import {types, removeNotification} from '../../src/store/notifications';
import {mockStore, checkPropTypes, genNotification} from '../utils/fixtures';

describe('<Notification/>', () => {
  let notification = null;
  let store = null;
  const className = theme.notification.className;
  // these props are the same for all tests
  const otherProps = {
    className,
    removeNotification
  };

  beforeEach(() => {
    notification = genNotification(1);
    // No image by default because newer versions of React do not add a trailing `;`
    // for inline style (in our case: `background-image`).
    notification.image = null;
    store = mockStore({notifications: []});
  });

  it('should validate props', () => {
    const errors = checkPropTypes({
      className,
      notification,
      removeNotification
    }, Notification.propTypes);
    expect(errors.className).toBeUndefined();
    expect(errors.notification).toBeUndefined();
    expect(errors.removeNotification).toBeUndefined();
  });

  it('should not validate props', () => {
    const errors = checkPropTypes({}, Notification.propTypes);
    expect(errors.className).toBeDefined();
    expect(errors.notification).toBeDefined();
    expect(errors.removeNotification).toBeDefined();
  });

  it('should mount with initial state', () => {
    // state component will be init without timer because `dismissAfter` is set to `0`
    notification.dismissAfter = 0;
    let wrapper = mount(<Notification notification={notification} {...otherProps}/>);
    expect(wrapper.state().timer).toEqual(null);
    // state component will be init with timer because `dismissAfter` > 0
    notification.dismissAfter = 1;
    wrapper = mount(<Notification notification={notification} {...otherProps}/>);
    expect(wrapper.state().timer).toBeInstanceOf(Timer);
  });

  it('should update state when receiving new props', () => {
    // state component will be init without timer because notification have buttons
    notification.dismissAfter = 0;
    let wrapper = mount(<Notification notification={notification} {...otherProps}/>);
    expect(wrapper.state().timer).toEqual(null);
    // we delete buttons to provoke creation of a Timer
    // at `componentWillReceivedProps()` component lifecycle
    notification.dismissAfter = 1;
    wrapper.setProps(notification);
    expect(wrapper.state().timer).toBeInstanceOf(Timer);
  });

  it('should render component (with title)', () => {
    const wrapper = shallow(
      <Notification notification={notification} {...otherProps}/>
    );
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render component (with HTML in the title)', () => {
    notification.title = 'A title with <i>html</i>';
    const wrapper = shallow(
      <Notification notification={notification} {...otherProps}/>
    );
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render component (without title)', () => {
    notification.title = null;
    const wrapper = shallow(
      <Notification notification={notification} {...otherProps}/>
    );
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render component (with message)', () => {
    const wrapper = shallow(
      <Notification notification={notification} {...otherProps}/>
    );
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render component (with HTML in the message)', () => {
    // add HTML in message and allow HTML
    notification.message = `${notification.message} <b>HEY</b>`;
    notification.allowHTML = true;
    const wrapper = shallow(
      <Notification notification={notification} {...otherProps}/>
    );
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render component (without message)', () => {
    notification.message = null;
    const wrapper = shallow(
      <Notification notification={notification} {...otherProps}/>
    );

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render component (with image)', () => {
    notification.image = 'an_url';
    const wrapper = shallow(
      <Notification notification={notification} {...otherProps}/>
    );
    // newer versions of React do not add a trailing `;`
    // for inline style (in our case: `background-image`), so we update the output
    // so the snapshots match when we run test with different versions of React.
    expect(wrapper.html().replace('an_url);', 'an_url)')).toMatchSnapshot();
  });

  it('should render component (without image)', () => {
    notification.image = null;
    const wrapper = shallow(
      <Notification notification={notification} {...otherProps}/>
    );

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render component (with close button)', () => {
    notification.buttons = [];
    notification.dismissible = true;
    notification.closeButton = true;
    const wrapper = shallow(
      <Notification notification={notification} {...otherProps}/>
    );

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render component (without close button)', () => {
    notification.buttons = [];
    notification.dismissible = true;
    notification.closeButton = false;
    const wrapper = shallow(
      <Notification notification={notification} {...otherProps}/>
    );

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render component (with 2 buttons)', () => {
    const wrapper = shallow(
      <Notification notification={notification} {...otherProps}/>
    );

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render component (with 1 button)', () => {
    delete notification.buttons[1];
    const wrapper = shallow(
      <Notification notification={notification} {...otherProps}/>
    );
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render component (without buttons)', () => {
    notification.buttons = [];
    const wrapper = shallow(
      <Notification notification={notification} {...otherProps}/>
    );

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should run onAdd() callback at componentDidMount() lifecycle', () => {
    let count = 0;
    // we throw an error to capture where
    // the code has been executed before error was thrown
    notification.onAdd = () => {
      count++;
    };

    mount(
      <Provider store={store}>
        <ConnectNotification notification={notification} {...otherProps}/>
      </Provider>
    );

    expect(count).toEqual(1);
  });

  it('should run onRemove() callback at componentWillUnmount() lifecycle', () => {
    let count = 0;
    notification.onRemove = () => {
      count++;
    };

    let wrapper = mount(
      <Provider store={store}>
        <ConnectNotification notification={notification} {...otherProps}/>
      </Provider>
    );

    wrapper.unmount();
    expect(count).toEqual(count);
  });

  it('should not throw an error at componentDidMount() lifecycle ' +
    '(onAdd() callback undefined)', () => {
    delete notification.onAdd;
    mount(
      <Provider store={store}>
        <ConnectNotification notification={notification} {...otherProps}/>
      </Provider>
    );
  });

  it('should not throw an error at componentWillUnmount() lifecycle ' +
    '(onRemove() callback undefined)', () => {
    delete notification.onRemove;
    const wrapper = mount(
      <Provider store={store}>
        <ConnectNotification notification={notification} {...otherProps}/>
      </Provider>
    );
    wrapper.unmount();
  });

  it('should create an action to remove the notification ' +
    'when it is clicked', () => {
    notification.dismissible = true;
    notification.buttons = [];
    notification.closeButton = false;

    const wrapper = mount(
      <Provider store={store}>
        <ConnectNotification notification={notification} {...otherProps}/>
      </Provider>
    );
    const expectedAction = {
      type: types.REMOVE_NOTIFICATION,
      payload: notification.id
    };

    wrapper.find(ConnectNotification).simulate('click');
    expect(store.getActions()).toEqual([expectedAction]);
  });

  it('should create an action to remove the notification ' +
    'when close button is clicked', () => {
    notification.dismissible = true;
    notification.buttons = [];
    notification.closeButton = true;
    const wrapper = mount(
      <Provider store={store}>
        <ConnectNotification notification={notification} {...otherProps}/>
      </Provider>
    );
    const expectedAction = {
      type: types.REMOVE_NOTIFICATION,
      payload: notification.id
    };
    wrapper.find(`.${className.closeButtonContainer} > span`).simulate('click');
    expect(store.getActions()).toEqual([expectedAction]);
  });

  it('should create an action to remove the notification ' +
    'when a action button is clicked (dismissible : false)', () => {
    // we set `dismissible` to `false` to be sure
    // that dismissible is ignored in this case
    notification.dismissible = false;
    const wrapper = mount(
      <Provider store={store}>
        <ConnectNotification notification={notification} {...otherProps}/>
      </Provider>
    );
    const expectedAction = {
      type: types.REMOVE_NOTIFICATION,
      payload: notification.id
    };

    wrapper.find(`.${className.button}`).first().simulate('click');
    expect(store.getActions()).toEqual([expectedAction]);
  });

  it('should create an action to remove the notification after ' +
    '`dismissAfter` duration', (done) => {
    const expectedAction = {
      type: types.REMOVE_NOTIFICATION,
      payload: notification.id
    };
    notification.dismissAfter = 10;
    // remove buttons otherwise `remove()` is not called after `dismissAfter` duration
    notification.buttons = [];
    mount(
      <Provider store={store}>
        <ConnectNotification notification={notification} {...otherProps}/>
      </Provider>
    );
    expect(store.getActions()).toEqual([]);
    setTimeout(() => {
      expect(store.getActions()).toEqual([expectedAction]);
      done();
    }, 15);
  });

  it('should not create an action to remove the notification ' +
    'while mouse is hovering it', (done) => {
    notification.dismissAfter = 100;
    notification.buttons = [];
    const wrapper = mount(
      <Provider store={store}>
        <ConnectNotification notification={notification} {...otherProps}/>
      </Provider>
    ).find(ConnectNotification);
    const expectedAction = {
      type: types.REMOVE_NOTIFICATION,
      payload: notification.id
    };
    // hover notification after 5s
    setTimeout(() => {
      wrapper.simulate('mouseEnter');
    }, 50);
    // check 5ms after `dismissAfter` duration that the store
    // is empty because mouse is hovering the notification
    setTimeout(() => {
      expect(store.getActions()).toEqual([]);
      // we leave notification
      wrapper.simulate('mouseLeave');
    }, 150);
    // and we check that the store contains an action
    setTimeout(() => {
      expect(store.getActions()).toEqual([expectedAction]);
      done();
    }, 210);
  });

  it('should not create an action to remove the notification ' +
    'when it is clicked (dismissible : false)', () => {
    notification.dismissible = false;
    const wrapper = mount(
      <Provider store={store}>
        <ConnectNotification notification={notification} {...otherProps}/>
      </Provider>
    );
    wrapper.find(ConnectNotification).simulate('click');
    expect(store.getActions()).toEqual([]);
  });

  it('should not create an action to remove the notification ' +
    'when a notification with a close button is clicked', () => {
    notification.dismissible = true;
    notification.buttons = [];
    notification.closeButton = true;
    const wrapper = mount(
      <Provider store={store}>
        <ConnectNotification notification={notification} {...otherProps}/>
      </Provider>
    );
    wrapper.find(ConnectNotification).simulate('click');
    expect(store.getActions()).toEqual([]);
  });

  it('should not create an action to remove the notification after ' +
    '`dismissAfter` duration (dismissAfter = 0)', (done) => {
    notification.dismissAfter = 0;
    mount(
      <Provider store={store}>
        <ConnectNotification notification={notification} {...otherProps}/>
      </Provider>
    );
    setTimeout(() => {
      expect(store.getActions()).toEqual([]);
      done();
    }, 10);
  });
});
