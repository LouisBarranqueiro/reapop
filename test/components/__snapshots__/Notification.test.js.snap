// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Notification/> should render component (with 1 button) 1`] = `"<div class=\\"notification-wrapper\\"><div class=\\"notification notification--default notification--buttons-2 \\"><span class=\\"fa notification-icon\\"></span><div class=\\"notification-meta\\"><h4 class=\\"notification-title\\">a title</h4><p class=\\"notification-message\\">a message</p></div><div class=\\"notification-close-button-container\\"><span class=\\"fa notification-close-button\\"></span></div><div class=\\"notification-buttons\\"><button class=\\"notification-button\\"><span class=\\"notification-button-text\\"><b>Yes</b></span></button></div></div></div>"`;

exports[`<Notification/> should render component (with 2 buttons) 1`] = `"<div class=\\"notification-wrapper\\"><div class=\\"notification notification--default notification--buttons-2 \\"><span class=\\"fa notification-icon\\"></span><div class=\\"notification-meta\\"><h4 class=\\"notification-title\\">a title</h4><p class=\\"notification-message\\">a message</p></div><div class=\\"notification-close-button-container\\"><span class=\\"fa notification-close-button\\"></span></div><div class=\\"notification-buttons\\"><button class=\\"notification-button\\"><span class=\\"notification-button-text\\"><b>Yes</b></span></button><button class=\\"notification-button\\"><span class=\\"notification-button-text\\">No</span></button></div></div></div>"`;

exports[`<Notification/> should render component (with HTML in the message) 1`] = `"<div class=\\"notification-wrapper\\"><div class=\\"notification notification--default notification--buttons-2 \\"><span class=\\"fa notification-icon\\"></span><div class=\\"notification-meta\\"><h4 class=\\"notification-title\\">a title</h4><p class=\\"notification-message\\">a message <b>HEY</b></p></div><div class=\\"notification-close-button-container\\"><span class=\\"fa notification-close-button\\"></span></div><div class=\\"notification-buttons\\"><button class=\\"notification-button\\"><span class=\\"notification-button-text\\"><b>Yes</b></span></button><button class=\\"notification-button\\"><span class=\\"notification-button-text\\">No</span></button></div></div></div>"`;

exports[`<Notification/> should render component (with HTML in the title) 1`] = `"<div class=\\"notification-wrapper\\"><div class=\\"notification notification--default notification--buttons-2 \\"><span class=\\"fa notification-icon\\"></span><div class=\\"notification-meta\\"><h4 class=\\"notification-title\\">A title with <i>html</i></h4><p class=\\"notification-message\\">a message</p></div><div class=\\"notification-close-button-container\\"><span class=\\"fa notification-close-button\\"></span></div><div class=\\"notification-buttons\\"><button class=\\"notification-button\\"><span class=\\"notification-button-text\\"><b>Yes</b></span></button><button class=\\"notification-button\\"><span class=\\"notification-button-text\\">No</span></button></div></div></div>"`;

exports[`<Notification/> should render component (with close button) 1`] = `"<div class=\\"notification-wrapper\\"><div class=\\"notification notification--default  \\"><span class=\\"fa notification-icon\\"></span><div class=\\"notification-meta\\"><h4 class=\\"notification-title\\">a title</h4><p class=\\"notification-message\\">a message</p></div><div class=\\"notification-close-button-container\\"><span class=\\"fa notification-close-button\\"></span></div></div></div>"`;

exports[`<Notification/> should render component (with image) 1`] = `"<div class=\\"notification-wrapper\\"><div class=\\"notification notification--default notification--buttons-2 \\"><div class=\\"notification-image-container\\"><span class=\\"notification-image\\" style=\\"background-image:url(an_url)\\"></span></div><div class=\\"notification-meta\\"><h4 class=\\"notification-title\\">a title</h4><p class=\\"notification-message\\">a message</p></div><div class=\\"notification-close-button-container\\"><span class=\\"fa notification-close-button\\"></span></div><div class=\\"notification-buttons\\"><button class=\\"notification-button\\"><span class=\\"notification-button-text\\"><b>Yes</b></span></button><button class=\\"notification-button\\"><span class=\\"notification-button-text\\">No</span></button></div></div></div>"`;

exports[`<Notification/> should render component (with message) 1`] = `"<div class=\\"notification-wrapper\\"><div class=\\"notification notification--default notification--buttons-2 \\"><span class=\\"fa notification-icon\\"></span><div class=\\"notification-meta\\"><h4 class=\\"notification-title\\">a title</h4><p class=\\"notification-message\\">a message</p></div><div class=\\"notification-close-button-container\\"><span class=\\"fa notification-close-button\\"></span></div><div class=\\"notification-buttons\\"><button class=\\"notification-button\\"><span class=\\"notification-button-text\\"><b>Yes</b></span></button><button class=\\"notification-button\\"><span class=\\"notification-button-text\\">No</span></button></div></div></div>"`;

exports[`<Notification/> should render component (with title) 1`] = `"<div class=\\"notification-wrapper\\"><div class=\\"notification notification--default notification--buttons-2 \\"><span class=\\"fa notification-icon\\"></span><div class=\\"notification-meta\\"><h4 class=\\"notification-title\\">a title</h4><p class=\\"notification-message\\">a message</p></div><div class=\\"notification-close-button-container\\"><span class=\\"fa notification-close-button\\"></span></div><div class=\\"notification-buttons\\"><button class=\\"notification-button\\"><span class=\\"notification-button-text\\"><b>Yes</b></span></button><button class=\\"notification-button\\"><span class=\\"notification-button-text\\">No</span></button></div></div></div>"`;

exports[`<Notification/> should render component (without buttons) 1`] = `"<div class=\\"notification-wrapper\\"><div class=\\"notification notification--default  \\"><span class=\\"fa notification-icon\\"></span><div class=\\"notification-meta\\"><h4 class=\\"notification-title\\">a title</h4><p class=\\"notification-message\\">a message</p></div><div class=\\"notification-close-button-container\\"><span class=\\"fa notification-close-button\\"></span></div></div></div>"`;

exports[`<Notification/> should render component (without close button) 1`] = `"<div class=\\"notification-wrapper\\"><div class=\\"notification notification--default  notification--dismissible\\"><span class=\\"fa notification-icon\\"></span><div class=\\"notification-meta\\"><h4 class=\\"notification-title\\">a title</h4><p class=\\"notification-message\\">a message</p></div></div></div>"`;

exports[`<Notification/> should render component (without image) 1`] = `"<div class=\\"notification-wrapper\\"><div class=\\"notification notification--default notification--buttons-2 \\"><span class=\\"fa notification-icon\\"></span><div class=\\"notification-meta\\"><h4 class=\\"notification-title\\">a title</h4><p class=\\"notification-message\\">a message</p></div><div class=\\"notification-close-button-container\\"><span class=\\"fa notification-close-button\\"></span></div><div class=\\"notification-buttons\\"><button class=\\"notification-button\\"><span class=\\"notification-button-text\\"><b>Yes</b></span></button><button class=\\"notification-button\\"><span class=\\"notification-button-text\\">No</span></button></div></div></div>"`;

exports[`<Notification/> should render component (without message) 1`] = `"<div class=\\"notification-wrapper\\"><div class=\\"notification notification--default notification--buttons-2 \\"><span class=\\"fa notification-icon\\"></span><div class=\\"notification-meta\\"><h4 class=\\"notification-title\\">a title</h4></div><div class=\\"notification-close-button-container\\"><span class=\\"fa notification-close-button\\"></span></div><div class=\\"notification-buttons\\"><button class=\\"notification-button\\"><span class=\\"notification-button-text\\"><b>Yes</b></span></button><button class=\\"notification-button\\"><span class=\\"notification-button-text\\">No</span></button></div></div></div>"`;

exports[`<Notification/> should render component (without title) 1`] = `"<div class=\\"notification-wrapper\\"><div class=\\"notification notification--default notification--buttons-2 \\"><span class=\\"fa notification-icon\\"></span><div class=\\"notification-meta\\"><p class=\\"notification-message\\">a message</p></div><div class=\\"notification-close-button-container\\"><span class=\\"fa notification-close-button\\"></span></div><div class=\\"notification-buttons\\"><button class=\\"notification-button\\"><span class=\\"notification-button-text\\"><b>Yes</b></span></button><button class=\\"notification-button\\"><span class=\\"notification-button-text\\">No</span></button></div></div></div>"`;
